homeassistant/components/recorder/history/modern.py

```diff
@@ -758,18 +758,30 @@ def _sorted_states_to_dict(
                 for row in group
                 if (state := row[state_idx]) != prev_state
             )
-            continue
+        else:
+            # Non-compressed state format returns an ISO formatted string
+            _utc_from_timestamp = dt_util.utc_from_timestamp
+            ent_results.extend(
+                {
+                    attr_state: (prev_state := state),  # noqa: F841
+                    attr_time: _utc_from_timestamp(
+                        row[last_updated_ts_idx]
+                    ).isoformat(),
+                }
+                for row in group
+                if (state := row[state_idx]) != prev_state
+            )
 
-        # Non-compressed state format returns an ISO formatted string
-        _utc_from_timestamp = dt_util.utc_from_timestamp
-        ent_results.extend(
-            {
-                attr_state: (prev_state := state),  # noqa: F841
-                attr_time: _utc_from_timestamp(row[last_updated_ts_idx]).isoformat(),
-            }
-            for row in group
-            if (state := row[state_idx]) != prev_state
-        )
+        # Some DB engines (seen on postgresql 14.8) will put the unioned
+        # start time state at the end of the list. We want it at the front.
+        if (
+            ent_results
+            and (last_result := ent_results[-1])
+            and type(last_result) is dict  # pylint: disable=unidiomatic-typecheck
+            and last_result[attr_time] is None
+        ):
+            last_result[attr_time] = start_time_ts
+            ent_results.insert(0, ent_results.pop())
```

Comment:
I'm not sure this code will actually ever be run on the CI because none of the db engines we are testing against have the problem

Reply:
Another option would be to resort the whole unioned query after its built but that would generate a temporary table and make everyone have to pay the performance hit for a postgresql only issue

This seems like the cheapest way to solve the problem without making everyone else pay the price